File: Finished/Remove Nth Node From End of List.go
package leetcode

import "leetcode/ds"

func removeNthFromEnd(head *ds.ListNode, n int) *ds.ListNode {
	if n == 0 {
		return head
	}

	length := helper2(head, n)
	if length == n {
		return head.Next
	}

	return head
}

func helper2(l *ds.ListNode, n int) int {
	if l == nil {
		return 0
	}

	length := 1 + helper2(l.Next, n)
	if length == n+1 {
		l.Next = l.Next.Next
	}
	return length
}

File: finished/remove_nth_node_from_end_of_list.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        def helper(node, n):
            if node == None:
                return 0
            
            length = helper(node.next, n) + 1
            
            if length == n + 1:
                node.next = node.next.next
            
            return length
        
        dummy_head = ListNode(None, head)

        helper(dummy_head, n)
        return dummy_head.next